File: src/algebra1_polynomial_operations_hard.py
"""
algebra1 - polynomial_operations (hard)
Generated: 2026-02-22T03:59:35.101981
"""

from problem_utils import *

def generate():
    problem_type = choice([
        'multiply_binomials_expanded',
        'factor_difference_of_squares',
        'multiply_three_binomials',
        'factor_perfect_square_trinomial',
        'polynomial_division',
        'multiply_with_distribution',
        'factor_by_grouping',
        'expand_binomial_power'
    ])
    
    if problem_type == 'multiply_binomials_expanded':
        # (ax + b)(cx + d)(ex + f) expanded form
        a_val = nonzero(-3, 3)
        b_val = nonzero(-5, 5)
        c_val = nonzero(-3, 3)
        d_val = nonzero(-5, 5)
        e_val = nonzero(-2, 2)
        f_val = nonzero(-5, 5)
        
        expr1 = a_val*x + b_val
        expr2 = c_val*x + d_val
        expr3 = e_val*x + f_val
        
        full_expr = expr1 * expr2 * expr3
        ans = expand(full_expr)
        
        return problem(
            question=f"Expand $({latex(expr1)})({latex(expr2)})({latex(expr3)})$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="algebra1/polynomial_operations",
            grading_mode="expand",
            solution=steps(
                f"First multiply $({latex(expr1)})({latex(expr2)}) = {latex(expand(expr1*expr2))}$",
                f"Then multiply by $({latex(expr3)})$",
                f"$({latex(expand(expr1*expr2))})({latex(expr3)}) = {latex(ans)}$"
            )
        )
    
    elif problem_type == 'factor_difference_of_squares':
        # Factor a^2x^2 - b^2y^2 type expression
        a_val = randint(2, 5)
        b_val = randint(2, 5)
        
        expr = (a_val*x)**2 - (b_val*y)**2
        ans = factor(expr)
        
        return problem(
            question=f"Factor ${latex(expr)}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra1/polynomial_operations",
            grading_mode="factor",
            solution=steps(
                f"Recognize difference of squares: $a^2 - b^2 = (a+b)(a-b)$",
                f"Here $a = {a_val}x$ and $b = {b_val}y$",
                f"${latex(ans)}$"
            )
        )
    
    elif problem_type == 'multiply_three_binomials':
        # (x + a)(x + b)(x + c) - find the coefficient of x^2 or x term
        a_val = nonzero(-6, 6)
        b_val = nonzero(-6, 6)
        c_val = nonzero(-6, 6)
        
        expr1 = x + a_val
        expr2 = x + b_val
        expr3 = x + c_val
        
        expanded = expand(expr1 * expr2 * expr3)
        coeff_x2 = expanded.as_coefficients_dict()[x**2]
        
        return problem(
            question=f"What is the coefficient of $x^2$ in the expansion of $({latex(expr1)})({latex(expr2)})({latex(expr3)})$?",
            answer=coeff_x2,
            difficulty=(1700, 1800),
            topic="algebra1/polynomial_operations",
            solution=steps(
                f"Expand $({latex(expr1)})({latex(expr2)})({latex(expr3)})$",
                f"Full expansion: ${latex(expanded)}$",
                f"Coefficient of $x^2$ is ${coeff_x2}$",
                f"Alternatively: coefficient of $x^2$ is $({a_val}) + ({b_val}) + ({c_val}) = {coeff_x2}$"
            )
        )
    
    elif problem_type == 'factor_perfect_square_trinomial':
        # Factor ax^2 + 2abx + b^2 or similar
        a_val = choice([2, 3, 4, 5])
        b_val = nonzero(-7, 7)
        
        expr = expand((a_val*x + b_val)**2)
        ans = factor(expr)
        
        return problem(
            question=f"Factor ${latex(expr)}$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="algebra1/polynomial_operations",
            grading_mode="factor",
            solution=steps(
                f"Recognize perfect square trinomial: $a^2 + 2ab + b^2 = (a+b)^2$",
                f"First term: $({a_val}x)^2$, last term: $({b_val})^2$",
                f"Middle term: $2 \\cdot {a_val}x \\cdot {b_val} = {2*a_val*b_val}x$",
                f"${latex(ans)}$"
            )
        )
    
    elif problem_type == 'polynomial_division':
        # (ax^3 + bx^2 + cx + d) / (x + e), find quotient
        e_val = nonzero(-5, 5)
        a_val = nonzero(-3, 3)
        b_val = randint(-10, 10)
        c_val = randint(-10, 10)
        
        # Make it divide evenly: construct dividend as (x + e) * (quadratic)
        q2 = a_val
        q1 = randint(-5, 5)
        q0 = randint(-5, 5)
        
        quotient = a_val*x**2 + q1*x + q0
        divisor = x + e_val
        dividend = expand(quotient * divisor)
        
        return problem(
            question=f"Divide ${latex(dividend)}$ by ${latex(divisor)}$. What is the quotient?",
            answer=quotient,
            difficulty=(1750, 1850),
            topic="algebra1/polynomial_operations",
            solution=steps(
                f"Use polynomial long division or synthetic division",
                f"Divide ${latex(dividend)}$ by ${latex(divisor)}$",
                f"Quotient: ${latex(quotient)}$"
            )
        )
    
    elif problem_type == 'multiply_with_distribution':
        # (ax^2 + bx + c)(dx + e)
        a_val = nonzero(-4, 4)
        b_val = nonzero(-6, 6)
        c_val = nonzero(-8, 8)
        d_val = nonzero(-3, 3)
        e_val = nonzero(-6, 6)
        
        expr1 = a_val*x**2 + b_val*x + c_val
        expr2 = d_val*x + e_val
        ans = expand(expr1 * expr2)
        
        return problem(
            question=f"Expand $({latex(expr1)})({latex(expr2)})$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra1/polynomial_operations",
            grading_mode="expand",
            solution=steps(
                f"Distribute each term of ${latex(expr2)}$ to each term of ${latex(expr1)}$",
                f"${latex(d_val*x)} \\cdot ({latex(expr1)}) = {latex(expand(d_val*x*expr1))}$",
                f"${latex(e_val)} \\cdot ({latex(expr1)}) = {latex(expand(e_val*expr1))}$",
                f"Add: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 'factor_by_grouping':
        # Factor ax^3 + bx^2 + cx + d by grouping
        # Construct as (mx + n)(px^2 + q)
        m_val = nonzero(-3, 3)
        n_val = nonzero(-5, 5)
        p_val = nonzero(-3, 3)
        q_val = nonzero(-5, 5)
        
        factor1 = m_val*x + n_val
        factor2 = p_val*x**2 + q_val
        expr = expand(factor1 * factor2)
        ans = factor(expr)
        
        return problem(
            question=f"Factor by grouping: ${latex(expr)}$",
            answer=ans,
            difficulty=(1750, 1850),
            topic="algebra1/polynomial_operations",
            grading_mode="factor",
            solution=steps(
                f"Group terms: $({latex(m_val*p_val*x**3 + m_val*q_val*x)}) + ({latex(n_val*p_val*x**2 + n_val*q_val)})$",
                f"Factor each group: ${latex(m_val*x)}({latex(p_val*x**2 + q_val)}) + {latex(n_val)}({latex(p_val*x**2 + q_val)})$",
                f"Common factor: $({latex(p_val*x**2 + q_val)})({latex(m_val*x + n_val)})$",
                f"${latex(ans)}$"
            )
        )
    
    else:  # expand_binomial_power
        # (ax + b)^3
        a_val = nonzero(-3, 3)
        b_val = nonzero(-4, 4)
        
        expr = (a_val*x + b_val)**3
        ans = expand(expr)
        
        return problem(
            question=f"Expand $({latex(a_val*x + b_val)})^3$",
            answer=ans,
            difficulty=(1700, 1800),
            topic="algebra1/polynomial_operations",
            grading_mode="expand",
            solution=steps(
                f"Use $(a+b)^3 = a^3 + 3a^2b + 3ab^2 + b^3$",
                f"Here $a = {a_val}x$ and $b = {b_val}$",
                f"$({a_val}x)^3 + 3({a_val}x)^2({b_val}) + 3({a_val}x)({b_val})^2 + ({b_val})^3$",
                f"${latex(ans)}$"
            )
        )

emit(generate())

File: src/algebra1_polynomial_operations_hard.jl
# algebra1 - polynomial_operations (hard)
# Generated: 2026-03-08T19:59:28.955055

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("algebra1/polynomial_operations")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Polynomial multiplication with higher degrees (1800-2100)
        deg1 = rand(2:4)
        deg2 = rand(2:3)
        
        p1 = rand_poly(x, deg1; coeff=(-8, 8))
        p2 = rand_poly(x, deg2; coeff=(-6, 6))
        
        product = expand(p1.expr * p2.expr)
        
        problem(
            question="Expand and simplify: \$($(tex(p1.expr)))($(tex(p2.expr)))\$",
            answer=product,
            difficulty=(1800, 2100),
            solution=steps(
                sol("Given product", p1.expr * p2.expr),
                "Apply distributive property to multiply each term",
                "Combine like terms",
                sol("Answer", product)
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Polynomial division with remainder (1900-2200)
        divisor_deg = rand(1:2)
        quotient_deg = rand(2:3)
        
        divisor_roots = [nonzero(-10, 10) for _ in 1:divisor_deg]
        divisor = prod(x - r for r in divisor_roots)
        divisor = expand(divisor)
        
        q = rand_poly(x, quotient_deg; coeff=(-7, 7))
        remainder = randint(-20, 20)
        
        dividend = expand(divisor * q.expr + remainder)
        
        problem(
            question="Divide \$$(tex(dividend))\$ by \$$(tex(divisor))\$ and express in the form \$Q(x) + \\frac{R}{$(tex(divisor))}\$ where \$R\$ is a constant.",
            answer="$(tex(q.expr)) + \\frac{$(remainder)}{$(tex(divisor))}",
            difficulty=(1900, 2200),
            grading_mode="symbolic",
            solution=steps(
                sol("Dividend", dividend),
                sol("Divisor", divisor),
                "Perform polynomial long division",
                sol("Quotient", q.expr),
                sol("Remainder", remainder),
                "Answer: \$$(tex(q.expr)) + \\frac{$(remainder)}{$(tex(divisor))}\$"
            ),
            time=240
        )
        
    elseif problem_type == 3
        # Composition of polynomials (2000-2300)
        inner = rand_quadratic(x; a=(-4, 4), b=(-6, 6), c=(-8, 8))
        outer_deg = rand(2:3)
        outer = rand_poly(x, outer_deg; coeff=(-5, 5))
        
        composition = substitute(outer.expr, x => inner.expr)
        composition = expand(composition)
        
        problem(
            question="Let \$f(x) = $(tex(outer.expr))\$ and \$g(x) = $(tex(inner.expr))\$. Find \$f(g(x))\$ in expanded form.",
            answer=composition,
            difficulty=(2000, 2300),
            solution=steps(
                sol("f(x)", outer.expr),
                sol("g(x)", inner.expr),
                "Substitute \$g(x)\$ into \$f(x)\$",
                sol("f(g(x))", substitute(outer.expr, x => inner.expr)),
                "Expand all terms",
                sol("Answer", composition)
            ),
            time=240
        )
        
    elseif problem_type == 4
        # Finding polynomial given roots and conditions (2100-2400)
        r1, r2, r3 = nonzero(-8, 8), nonzero(-8, 8), nonzero(-8, 8)
        while r1 == r2 || r2 == r3 || r1 == r3
            r1, r2, r3 = nonzero(-8, 8), nonzero(-8, 8), nonzero(-8, 8)
        end
        
        point_x = nonzero(-6, 6)
        while point_x in [r1, r2, r3]
            point_x = nonzero(-6, 6)
        end
        point_y = nonzero(-30, 30)
        
        base_poly = expand((x - r1) * (x - r2) * (x - r3))
        eval_base = substitute(base_poly, x => point_x)
        
        a_coeff = point_y // eval_base
        answer_poly = expand(simplify(a_coeff * base_poly))
        
        problem(
            question="Find the cubic polynomial \$P(x)\$ with roots at \$x = $(r1), $(r2), $(r3)\$ such that \$P($(point_x)) = $(point_y)\$. Express in standard form.",
            answer=answer_poly,
            difficulty=(2100, 2400),
            solution=steps(
                "Start with \$P(x) = a(x - $(r1))(x - $(r2))(x - $(r3))\$",
                sol("Base polynomial", base_poly),
                "Use condition \$P($(point_x)) = $(point_y)\$",
                "Calculate \$a = \\frac{$(point_y)}{$(eval_base)} = $(a_coeff)\$",
                sol("Answer", answer_poly)
            ),
            time=270
        )
        
    else
        # Polynomial with parametric coefficient (2200-2500)
        @variables k
        
        p1 = rand_quadratic(x; a=(-5, 5), b=(-8, 8), c=(-10, 10))
        p2 = rand_linear(x; a=(-6, 6), b=(-8, 8))
        
        result = expand(p1.expr + k * p2.expr)
        target_coeff_power = 1
        
        coeff_of_x = p1.b + k * p2.a
        k_value = solve(coeff_of_x ~ 0, k)[1]
        
        final_poly = substitute(result, k => k_value)
        
        problem(
            question="Consider the polynomial \$P(x) = $(tex(p1.expr)) + k($(tex(p2.expr)))\$ where \$k\$ is a real parameter. Find the value of \$k\$ such that the coefficient of \$x\$ in \$P(x)\$ is zero, and determine the resulting polynomial.",
            answer=final_poly,
            difficulty=(2200, 2500),
            solution=steps(
                sol("Given polynomial", result),
                "Identify coefficient of \$x\$: \$$(tex(coeff_of_x))\$",
                "Set coefficient equal to zero and solve for \$k\$",
                sol("k", k_value),
                "Substitute back into polynomial",
                sol("Answer", final_poly)
            ),
            time=300
        )
    end
end